File: BOJ(Java 공부용)/10825.java
import java.io.*;
import java.util.ArrayList;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().split(" ")[0]);
        ArrayList<p> v = new ArrayList<>();
        for(int i=0;i<n;i++) {
            String[] input = br.readLine().split(" ");
            int kor = Integer.parseInt(input[1]);
            int eng = Integer.parseInt(input[2]);
            int math = Integer.parseInt(input[3]);
            v.add(new p(kor,math,eng,input[0]));
        }
        v.sort((a, b) -> {
            if(a.kor > b.kor) return -1;
            else if(a.kor < b.kor) return 1;
            else {
                if(a.eng<b.eng) return -1;
                else if(a.eng>b.eng) return 1;
                else {
                    if(a.math>b.math) return -1;
                    else if(a.math<b.math) return 1;
                    else {
                        return a.name.compareTo(b.name);
                    }
                }
            }
        });
        for(int i=0;i<v.size();i++) {
            bw.write(v.get(i).name+"\n");
        }
        bw.flush();
        bw.close();
    }
    private static class p{
        int kor,math,eng;
        String name;
        private p(int kor,int math,int eng,String name) {
            this.kor = kor;
            this.math = math;
            this.eng = eng;
            this.name = name;
        }
    }
}

File: BOJ(Kotlin)/10825.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val n = br.readLine().toInt()
    val arr = mutableListOf<Pos>()
    for (i in 0 until n) {
        val st = StringTokenizer(br.readLine())
        val name = st.nextToken()
        val kor = st.nextToken().toInt()
        val eng = st.nextToken().toInt()
        val math = st.nextToken().toInt()
        arr.add(Pos(kor,eng,math,name))
    }

    arr.sortWith {a,b ->
        when {
            a.kor > b.kor -> -1
            a.kor < b.kor -> 1
            else -> {
                when {
                    a.eng < b.eng -> -1
                    a.eng > b.eng -> 1
                    else -> {
                        when {
                            a.math > b.math -> -1
                            a.math < b.math -> 1
                            else -> a.name.compareTo(b.name)
                        }
                    }
                }
            }
        }
    }

    for (value in arr) {
        bw.write(value.name+"\n")
    }
    bw.flush()
    bw.close()
}
// parameter에 val을 붙이고 안붙이고 무슨 차이지?
class Pos(val kor : Int, val eng: Int, val math: Int, val name : String){}